lib/rspec/rails/matchers/have_reported_error.rb

```diff
@@ -0,0 +1,271 @@
+            end
+          end
+        end
+
+        def unmatched_attributes(actual)
+          @attributes.reject do |key, value|
+            if value.respond_to?(:matches?)
+              value.matches?(actual[key])
+            else
+              actual[key] == value
+            end
+          end
+        end
+
+        def warn_about_nil_error!
+          RSpec.warn_with("Using the `have_reported_error` matcher with a `nil` error is probably " \
+                         "unintentional, it risks false positives, since `have_reported_error` " \
+                         "will match when any error is reported to Rails.error, potentially " \
+                         "allowing the expectation to pass without the specific error you are " \
+                         "intending to test for being reported. " \
+                         "Instead consider providing a specific error class or message. " \
+                         "This message can be suppressed by setting: " \
+                         "`RSpec::Expectations.configuration.on_potential_false" \
+                         "_positives = :nothing`")
+        end
+      end
+
+      # @api public
+      # Passes if the block reports an error to `Rails.error`.
+      #
+      # This matcher asserts that ActiveSupport::ErrorReporter has received an error report.
+      #
+      # @example Checking for any error
+      #   expect { Rails.error.report(StandardError.new) }.to have_reported_error
+      #
+      # @example Checking for specific error class
+      #   expect { Rails.error.report(MyError.new) }.to have_reported_error(MyError)
+      #
+      # @example Checking for specific error class with message
+      #   expect { Rails.error.report(MyError.new("message")) }.to have_reported_error(MyError, "message")
+      #
+      # @example Checking for error with exact message (any class)
+      #   expect { Rails.error.report(StandardError.new("exact message")) }.to have_reported_error("exact message")
+      #
+      # @example Checking for error with message pattern (any class)
+      #   expect { Rails.error.report(StandardError.new("test message")) }.to have_reported_error(/test/)
+      #
+      # @example Checking for specific error class with message pattern
+      #   expect { Rails.error.report(StandardError.new("test message")) }.to have_reported_error(StandardError, /test/)
+      #
+      # @example Checking error attributes
+      #   expect { Rails.error.report(StandardError.new, context: "test") }.to have_reported_error.with_context(context: "test")
+      #
+      # @example Negation
+      #   expect { "safe code" }.not_to have_reported_error
+      #
+      # @param expected_error_or_message [Class, String, Regexp, nil] the expected error class, message string, or message pattern
+      # @param expected_message [String, Regexp, nil] the expected error message to match
+      def have_reported_error(expected_error_or_message = UndefinedValue, expected_message = nil)
```

Comment:
We have default attributes both here and in the initializer. Worth leaving just here?

Reply:
Resolved!

Reply:
There are still defaults both here and [there](https://github.com/rspec/rspec-rails/pull/2849/files/c1816e2f2b14a547e26ddfc0b2b3d9a02c4b9286#diff-5095b0a095ba1b0511036361c8930a18ff4367591813576e280b69bb823bf655R38):
```
def initialize(expected_error_or_message = UndefinedValue, expected_message = nil)
```
Let's remove there.